File: Leetcode/src/com/leetcode/bitManipulation/_137_SingleNumberII.java
package com.leetcode.bitManipulation;

/**
 * @Date: 07/18/2020
 * @Description: bit,
 **/
public class _137_SingleNumberII {

    // 主要是看每一位的情况
    //time:O(n) space:O(1)
    public int singleNumber(int[] nums) {
        int res = 0;
        if (nums == null || nums.length == 0) {
            return res;
        }
        // 计算每一位上的1's的个数
        for (int i = 0; i < 32; i++) {
            int sum = 0;
            for (int j = 0; j < nums.length; j++) {
                if (((nums[j] >> i) & 1) == 1) {
                    sum++;
                }
            }
            sum %= 3;
            if (sum != 0) {
                res |= (sum << i);
            }
        }
        return res;
    }
}

File: Leetcode/src/com/leetcode/bit_manipulation/_137_single_number_ii.py
from typing import List


class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        if not nums:
            return 0

        res = 0
        for i in range(32):
            total = 0
            for j in range(len(nums)):
                if (nums[j] >> i) & 1 == 1:
                    total += 1
            total %= 3
            if total != 0:
                res |= total << i
        return res if res < 2**31 else res - 2**32
